### PR TITLE
[SYCL-MLIR] Call cgeist from clang

### PR DESCRIPTION
**Pull request by Naghasan**
When appending `-syclmlir` to the target triple, the driver will invoke cgeist instead of clang.
The default setup is to call cgeist to emit LLVM IR. This allow the insertion into the SYCL compilation flow.
To produce an mlir output, clang can be invoked with `-fsycl-device-only -emit-mlir`.

@etiotto @whitneywhtsang I picked the triple approach to avoid adding a flag to ease up integration overall, that should help with sycl tests later on (as you have to pass a triple).

Signed-off-by: Victor Lomuller <victor@codeplay.com>

### CODE REVIEW

File: clang/lib/Driver/ToolChains/Clang.h [hunk 1]
```diff
@@ -309,6 +309,28 @@ class LLVM_LIBRARY_VISIBILITY LinkerWrapper final : public Tool {
                     const char *LinkingOutput) const override;
 };
 
+/// Offload bundler tool.
```
Comment (asudarsa) [inline clang/lib/Driver/ToolChains/Clang.h:312]: Minor nit - Incorrect comment here.

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 2]
```diff
@@ -9912,3 +9913,53 @@ void LinkerWrapper::ConstructJob(Compilation &C, const JobAction &JA,
...
   LinkCommand->replaceArguments(CmdArgs);
 }
+
+// Build an invocation to cgeist by stealing a clang invocation.
+// TODO: cgeist's driver actually replicate large part of clang.
+//       This make sense given the object and structure of the upstream project
+//       but given our context, a clang plugin would do just fine. We simply
+//       need to provide the frontend action (which already exist) via the
+//       plugin interface. As we control clang, we can make driver adjustments
+//       where needed.
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/Clang.cpp:9923]:
```suggestion
// TODO: cgeist's driver actually replicates large part of clang.
//       This makes sense given the object and structure of the upstream project
//       but given our context, a clang plugin would do just fine. We simply
//       need to provide the frontend action (which already exists) via the
//       plugin interface. As we control clang, we can make driver adjustments
//       where needed.
```

File: clang/lib/Driver/Driver.cpp [hunk 3]
```diff
@@ -4786,11 +4786,20 @@ class OffloadingActionBuilder final {
...
           types::ID OutputType = types::TY_LLVM_BC;
           if ((SYCLDeviceOnly || Args.hasArg(options::OPT_emit_llvm)) &&
               Args.hasArg(options::OPT_S))
             OutputType = types::TY_LLVM_IR;
+          if (SYCLDeviceOnly &&
+              TargetInfo.TC->getTriple().getEnvironment() ==
+                  llvm::Triple::SYCLMLIR &&
+              Args.hasArg(options::OPT_emit_mlir)) {
+            OutputType = types::TY_MLIR_IR;
+          }
```
Comment (etiotto) [inline clang/lib/Driver/Driver.cpp:4802]: If a source file is compiled using options `-fsycl-device-only -emit-mlir` we should emit MLIR code. Why do we also need to check that the triple is  'SYCLMLIR' ?

**Comment by etiotto** [comment]
@Naghasan I got your branch code but I am having problems. The `clang` driver does not accepts the `-emit-mlir` option.
Given `-fsycl-device-only -S -emit-llvm` clang emits the LLVM IR for the device:
```
 ~/projects/intel-llvm/build  victor/clang-mlir  ~/projects/intel-llvm/build/bin/clang ~/projects/tmp/sycl_single.cpp -fsycl- device-only -S -emit-llvm        
 ~/projects/intel-llvm/build  victor/clang-mlir  ls *.ll   
sycl_single-sycl-spir64-unknown-unknown.ll
```
However given `-fsycl-device-only -S -emit-mlir` I got an error:
```
 ~/projects/intel-llvm/build  victor/clang-mlir  ~/projects/intel-llvm/build/bin/clang ~/projects/tmp/sycl_single.cpp -fsycl-device-only -S -emit-mlir                
clang-16: warning: argument unused during compilation: '-emit-mlir' [-Wunused-command-line-argument]
```

Note that `-emit-mlir` is listed as a valid option:
```
 ~/projects/intel-llvm/build  victor/clang-mlir  ~/projects/intel-llvm/build/bin/clang --help|grep emit-mlir                                                      
  -emit-mlir              Build the parse tree, then lower it to MLIR
```

I think it would make sense for clang to accept that option, call cgeist and dump the MLIR produced by cgeist (that is the lowered MLIR which is the MLIR LLVM dialect) to a  sycl_single.mlir. What do you think?

**Comment by etiotto** [comment]
I think it would be nice if we had separate invocation commands (e.g. clang-mlir, clang-mlir++) that we could use rather than a special triple. I see that `clang-16` can differentiate between clang/clang++ invocations (clang and clang++ are symlinks to clang-16): 

```
 ~/projects/intel-llvm/build  victor/clang-mlir  ./bin/clang t.cpp -c                                                                                                 
 ~/projects/intel-llvm/build  victor/clang-mlir  ./bin/clang++ t.c -c                                                                                                   
clang-16: warning: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated [-Wdeprecated]
``` 

Can we do something similar ?

File: clang/lib/Driver/Driver.cpp (hunk 3 above)
Comment (Naghasan) [reply clang/lib/Driver/Driver.cpp]: Fixed

**Comment by Naghasan** [comment]
> 
> However given `-fsycl-device-only -S -emit-mlir` I got an error:
> 
> ```
>  ~/projects/intel-llvm/build  victor/clang-mlir  ~/projects/intel-llvm/build/bin/clang ~/projects/tmp/sycl_single.cpp -fsycl-device-only -S -emit-mlir                
> clang-16: warning: argument unused during compilation: '-emit-mlir' [-Wunused-command-line-argument]
> ```
> 

You forgot `-fsycl` :)


> I think it would be nice if we had separate invocation commands (e.g. clang-mlir, clang-mlir++) that we could use rather than a special triple. I see that `clang-16` can differentiate between clang/clang++ invocations (clang and clang++ are symlinks to clang-16):
> 
> ```
>  ~/projects/intel-llvm/build  victor/clang-mlir  ./bin/clang t.cpp -c                                                                                                 
>  ~/projects/intel-llvm/build  victor/clang-mlir  ./bin/clang++ t.c -c                                                                                                   
> clang-16: warning: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated [-Wdeprecated]
> ```
> 
> Can we do something similar ?

We could. We can introduce a driver mode implicitly setting `-fsycl -fsycl-device-only -emit-mlir` or other. But better to do that in another PR after we discuss what we want here.

The rational for the triple is integrate into the whole compilation logic of SYCL, by default the driver sets cgeist to output llvm ir so that it can be used transparently by the rest of the TC. The triple is the simplest to do this for now (tool selection is trivial this way). I agree it is a bit of a pain, but that make more sense when it comes to run SYCL tests / integrating into existing projects.

I that picture, `-emit-mlir` is actually just a way to introspect the intermediate result.

To save some typing, I could also introduce a short-hand like `spir64-mlir` to `-fsycl-targets`.

**Comment by etiotto** [comment]
I am not keen in using the triple to select whether `clang` translate its AST to LLVM IR or MLIR. The triple is used by the backend (`opt` and `llc`) an those tools will not know what to do with a new triple.  I would rather use a clang option to select whether `clang-16` lowers the AST is to LLVM IR or MLIR. 

In the end I think we will have a different compilation command  `clang-mlir` (a symlink to `clang-16`),  so `clang-16` will be able to detect the name of the symlink and use that name to call cgeist (no need to add any new option to clang-16).

When we want to run a SYCL test (a .cpp file) we can simply invoke the `clang-mlir` command.

**Review by etiotto — APPROVED** [review]
LGTM

**Pull request merged**